CryptoAnalysis/src/main/java/crypto/rules/CryptSLRuleReader.java

```diff
@@ -1,22 +1,154 @@
 package crypto.rules;
 
 import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
 import java.net.MalformedURLException;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Enumeration;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipFile;
 
 import crypto.cryptslhandler.CrySLModelReader;
 
 public class CryptSLRuleReader {
+
 	private static CrySLModelReader csmr;
-	
-	public static CryptSLRule readFromSourceFile(File file) {
-		if(csmr == null)
+
+	private static CrySLModelReader getReader(){
+		if (csmr == null)
+		{
 			try {
 				csmr = new CrySLModelReader();
-			} catch (MalformedURLException e) {
-				// TODO Auto-generated catch block
+			}
+			catch (MalformedURLException e){
 				e.printStackTrace();
+				// Sebastian:
+				// TODO: Current code could cause a NullPointerException
```

Comment:
Yes and no. If people write rules in the CrySL editor there is no chance of this exception being thrown. That said, I know of more people who have written rules without the editor than even know that it exists.